---
title: "Example workflow qPCRanalysis"
author: "Jordi Camps"
date: "26 February 2019"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Load packages
```{r message=FALSE}
library(readxl)
library(tidyr)
library(dplyr)
library(ggplot2)
library(qPCRanalysis)
library(ggpubr)
library(lazyeval)
```

## Load data
We load the data with the **read_excel()** function of the readxl package. The data should consist of at least 4 columns:

* __Sample__ : invidividual names per sample
* __Variable__ : the biological variability in your data set like a disease, different timepoints, perturbation ...
* __Target__ : the gene names of your primers
* __CT__ : the raw CT values

```{r message=FALSE, warning=FALSE}
qpcr <- read_excel("qPCR_data.xlsx", col_names = TRUE)
qpcr$CT <- as.numeric(qpcr$CT)
head(qpcr)
```

## Calculate Delta CT
To calculate delta CT we use the `calculate_DCT()` function. This function requires four arguments:

* __df__ : dataframe structured like the proposed data file
* __hkg__ : name of housekeeping gene or genes that you want to use to normalize against
* __sample_col__ : name of the sample column
* __gene_col__ : name of the gene column

It will pass a dataframe with three added columns: 

* __CT_hkg__ : average CT value of housekeeping genes
* __DCT__ : Delta CT values
* __RE__ : relative expression to hkg

```{r}
qpcr <- calculate_DCT(df = qpcr, hkg = c("Rab35", "Rpl13a", "PSma3"), sample_col = "Sample", gene_col = "Target")
```

## Calculate Delta Delta CT
To calculate Delta Delta CT use the `calculate_DDCT()` function. This function  can only be run after the `calculate_DCT()` function is used and requires five argeuments:

* __df__: dataframe structured like the proposed data file, has to contain DCT and RE column
* __gene_col__ : name of the gene column
* __sample_col__ : name of the sample column
* __var_col__ : column name of variables to normalize your control against
* __control__: name of variable to use as control

It will pass a dataframe with seven added columns

* __DCTavg__ : average Delta CT value
* __DCTsem__ : s.e.m. of Delta CT
* __DCTsemperc__ : percentage of s.e.m.
* __DDCTavg__ : average Delta Delta CT value
* __DDCTsem__ : standard error to the mean of Delta Delta CT
* __DDCTmin__ : minimum sem value
* __DDCTmax__ : maximum sem value

```{r}
qpcr <- calculate_DDCT(df = qpcr, gene_col = "Target", sample_col = "Sample", var = "Genotype", control = "Healthy")
head(qpcr)
```

